pkg/graveler/graveler.go

```diff
@@ -1219,57 +1221,192 @@ func (g *KVGraveler) DeleteBranch(ctx context.Context, repositoryID RepositoryID
 
 func (g *KVGraveler) SetGarbageCollectionRules(ctx context.Context, repositoryID RepositoryID, rules *GarbageCollectionRules) error {
-	return g.db.SetGarbageCollectionRules(ctx, repositoryID, rules)
+	repo, err := g.RefManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return err
+	}
+	return g.garbageCollectionManager.SaveRules(ctx, repo.StorageNamespace, rules)
 }
 
 func (g *KVGraveler) SaveGarbageCollectionCommits(ctx context.Context, repositoryID RepositoryID, previousRunID string) (*GarbageCollectionRunMetadata, error) {
-	return g.db.SaveGarbageCollectionCommits(ctx, repositoryID, previousRunID)
+	rules, err := g.GetGarbageCollectionRules(ctx, repositoryID)
+	if err != nil {
+		return nil, fmt.Errorf("get gc rules: %w", err)
+	}
+	repo, err := g.RefManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return nil, fmt.Errorf("get repository: %w", err)
+	}
+	previouslyExpiredCommits, err := g.garbageCollectionManager.GetRunExpiredCommits(ctx, repo.StorageNamespace, previousRunID)
+	if err != nil {
+		return nil, fmt.Errorf("get expired commits from previous run: %w", err)
+	}
+
+	runID, err := g.garbageCollectionManager.SaveGarbageCollectionCommits(ctx, repo.StorageNamespace, repositoryID, rules, previouslyExpiredCommits)
+	if err != nil {
+		return nil, fmt.Errorf("save garbage collection commits: %w", err)
+	}
+	commitsLocation, err := g.garbageCollectionManager.GetCommitsCSVLocation(runID, repo.StorageNamespace)
+	if err != nil {
+		return nil, err
+	}
+	addressLocation, err := g.garbageCollectionManager.GetAddressesLocation(repo.StorageNamespace)
+	if err != nil {
+		return nil, err
+	}
+
+	return &GarbageCollectionRunMetadata{
+		RunId:              runID,
+		CommitsCsvLocation: commitsLocation,
+		AddressLocation:    addressLocation,
+	}, err
 }
 
 func (g *KVGraveler) GetBranchProtectionRules(ctx context.Context, repositoryID RepositoryID) (*BranchProtectionRules, error) {
-	return g.db.GetBranchProtectionRules(ctx, repositoryID)
+	return g.protectedBranchesManager.GetRules(ctx, repositoryID)
 }
 
 func (g *KVGraveler) DeleteBranchProtectionRule(ctx context.Context, repositoryID RepositoryID, pattern string) error {
-	return g.db.DeleteBranchProtectionRule(ctx, repositoryID, pattern)
+	return g.protectedBranchesManager.Delete(ctx, repositoryID, pattern)
 }
 
 func (g *KVGraveler) CreateBranchProtectionRule(ctx context.Context, repositoryID RepositoryID, pattern string, blockedActions []BranchProtectionBlockedAction) error {
-	return g.db.CreateBranchProtectionRule(ctx, repositoryID, pattern, blockedActions)
+	return g.protectedBranchesManager.Add(ctx, repositoryID, pattern, blockedActions)
+}
+
+// getFromStagingArea returns the most updated value of a given key in a branch staging area.
+// Iterate over all tokens - staging + sealed in order of last modified. First appearance of key represents the latest update
+func (g *KVGraveler) getFromStagingArea(ctx context.Context, b *Branch, key Key) (*Value, error) {
```

Comment:
How can we use concurrency in this function? The order matters here

Reply:
Right - the logic MUST remain as it is. But fetching the values serially is costly in time. You can try to fetch them all together and then apply the logic (we don't even have to wait for all requests if the value was found). However, it is costly in number of requests, since some tokens wiil be read unnecessarily.

Reply:
I have some reservations about this approach. For starters, The size of the token list should be relatively small in any normal scenario. If we reached a very big list - something really bad happened.
Secondly, if we fetch the values together, we will have a problem finding out the order of the values. And lastly as you stated this can potentially generate a lot of redundant calls. Lets discuss it further if you think it's important, and we can document it in an issue if needed